// vtex/lib/clients/sponsor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const api_1 = require("@vtex/api");
class Sponsor extends api_1.IOClient {
    constructor(context, options) {
        super(context, Object.assign(Object.assign({}, options), { authType: api_1.AuthType.bearer }));
        this.getSponsorAccount = () => __awaiter(this, void 0, void 0, function* () { return this.http.get(this.routes.getSponsorAccount, { metric: 'get-sponsor-account' }); });
        this.getEdition = () => __awaiter(this, void 0, void 0, function* () { return this.http.get(this.routes.getEdition, { metric: 'get-edition' }); });
        this.setEdition = (account, editionApp) => __awaiter(this, void 0, void 0, function* () {
            const [edition, version] = editionApp.split('@');
            const [sponsor, editionName] = edition.split('.');
            return this.http.post(this.routes.setEdition(account), { sponsor, edition: editionName, version }, { metric: 'set-edition' });
        });
        this.runHouseKeeper = () => __awaiter(this, void 0, void 0, function* () { return this.http.post(this.routes.runHouseKeeper, {}, { metric: 'run-house-keeper' }); });
        const { region, account, workspace } = context;
        this.region = region;
        this.account = account;
        this.workspace = workspace;
    }
    get routes() {
        return {
            getSponsorAccount: `http://kube-router.${this.region}.vtex.io/_account/${this.account}`,
            getEdition: `http://apps.${this.region}.vtex.io/${this.account}/${this.workspace}/edition`,
            setEdition: (account) => `http://tenant-provisioner.vtex.${this.region}.vtex.io/${this.account}/master/tenants/${account}/migrate`,
            runHouseKeeper: `http://housekeeper.${this.region}.vtex.io/${this.account}/master/_housekeeping/perform`,
        };
    }
}
exports.Sponsor = Sponsor;
